File: Applications/Examples/EDP/python/market_price_edpgw_authentication.py
#!/usr/bin/env python
#|-----------------------------------------------------------------------------
#|            This source code is provided under the Apache 2.0 license      --
#|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
#|                See the project's LICENSE.md for details.                  --
#|           Copyright Thomson Reuters 2018. All rights reserved.            --
#|-----------------------------------------------------------------------------
"""
Simple example of authenticating to EDP-GW and using the token to login and
retrieve MarketPrice content.  A username and password are used to retrieve
this token.
"""

import sys
import time
import getopt
import requests
import socket
import json
import websocket
import threading

# Global Default Variables
app_id = '256'
auth_url = 'https://api.refinitiv.com:443/auth/oauth2/beta1/token'
hostname = ''
password = ''
position = ''
sts_token = ''
refresh_token = ''
user = ''
clientid = ''
port = '443'
client_secret = ''
scope = 'trapi'
ric = '/TRI.N'
service = 'ELEKTRON_DD'
view = []
no_streaming = False

# Global Variables
web_socket_app = None
web_socket_open = False
logged_in = False


def process_message(message_json):
    """ Parse at high level and output JSON of message """
    message_type = message_json['Type']

    if message_type == "Refresh":
        if 'Domain' in message_json:
            message_domain = message_json['Domain']
            if message_domain == "Login":
                process_login_response(message_json)
    elif message_type == "Ping":
        pong_json = {'Type': 'Pong'}
        web_socket_app.send(json.dumps(pong_json))
        print("SENT:")
        print(json.dumps(pong_json, sort_keys=True, indent=2, separators=(',', ':')))


def process_login_response(message_json):
    """ Send item request """
    global logged_in

    if message_json['State']['Stream'] != "Open" or message_json['State']['Data'] != "Ok":
        print("Login failed.")
        sys.exit(1)

    logged_in = True
    send_market_price_request(ric)


def send_market_price_request(ric_name):
    """ Create and send simple Market Price request """
    mp_req_json = {
        'ID': 2,
        'Key': {
            'Name': ric_name,
            'Service': service
        },
        'Streaming': not no_streaming,
    }

    if view:
        mp_req_json['View'] = view

    web_socket_app.send(json.dumps(mp_req_json))
    print("SENT:")
    print(json.dumps(mp_req_json, sort_keys=True, indent=2, separators=(',', ':')))


def send_login_request(auth_token, is_refresh_token):
    """
        Send login request with authentication token.
        Used both for the initial login and subsequent reissues to update the authentication token
    """
    login_json = {
        'ID': 1,
        'Domain': 'Login',
        'Key': {
            'NameType': 'AuthnToken',
            'Elements': {
                'ApplicationId': '',
                'Position': '',
                'AuthenticationToken': ''
            }
        }
    }

    login_json['Key']['Elements']['ApplicationId'] = app_id
    login_json['Key']['Elements']['Position'] = position
    login_json['Key']['Elements']['AuthenticationToken'] = auth_token

    # If the token is a refresh token, this is not our first login attempt.
    if is_refresh_token:
        login_json['Refresh'] = False

    web_socket_app.send(json.dumps(login_json))
    print("SENT:")
    print(json.dumps(login_json, sort_keys=True, indent=2, separators=(',', ':')))


def on_message(_, message):
    """ Called when message received, parse message into JSON for processing """
    print("RECEIVED: ")
    message_json = json.loads(message)
    print(json.dumps(message_json, sort_keys=True, indent=2, separators=(',', ':')))

    for singleMsg in message_json:
        process_message(singleMsg)


def on_error(_, error):
    """ Called when websocket error has occurred """
    print(error)


def on_close(_):
    """ Called when websocket is closed """
    global web_socket_open
    web_socket_open = False
    print("WebSocket Closed")


def on_open(_):
    """ Called when handshake is complete and websocket is open, send login """

    print("WebSocket successfully connected!")
    global web_socket_open
    web_socket_open = True
    send_login_request(sts_token, False)


def get_sts_token(current_refresh_token, url=None):
    """
        Retrieves an authentication token.
        :param current_refresh_token: Refresh token retrieved from a previous authentication, used to retrieve a
        subsequent access token. If not provided (i.e. on the initial authentication), the password is used.
    """

    if url is None:
        url = auth_url

    if not current_refresh_token:  # First time through, send password
        if url.startswith('https'):
            data = {'username': user, 'password': password, 'grant_type': 'password', 'takeExclusiveSignOnControl': True,
                    'scope': scope}
        else:
            data = {'username': user, 'password': password, 'client_id': clientid, 'grant_type': 'password', 'takeExclusiveSignOnControl': True,
                    'scope': scope}
        print("Sending authentication request with password to", url, "...")
    else:  # Use the given refresh token
        if url.startswith('https'):
            data = {'username': user, 'refresh_token': current_refresh_token, 'grant_type': 'refresh_token'}
        else:
            data = {'username': user, 'client_id': clientid, 'refresh_token': current_refresh_token, 'grant_type': 'refresh_token'}
        print("Sending authentication request with refresh token to", url, "...")

    try:
        if url.startswith('https'):
            # Request with auth for https protocol
            r = requests.post(url,
                              headers={'Accept': 'application/json'},
                              data=data,
                              auth=(clientid, client_secret),
                              verify=True,
                              allow_redirects=False)
        else:
            # Request without auth for non https protocol (e.g. http)
            r = requests.post(url,
                              headers={'Accept': 'application/json'},
                              data=data,
                              verify=True,
                              allow_redirects=False)

    except requests.exceptions.RequestException as e:
        print('EDP-GW authentication exception failure:', e)
        return None, None, None

    if r.status_code == 200:
        auth_json = r.json()
        print("EDP-GW Authentication succeeded. RECEIVED:")
        print(json.dumps(auth_json, sort_keys=True, indent=2, separators=(',', ':')))

        return auth_json['access_token'], auth_json['refresh_token'], auth_json['expires_in']
    elif r.status_code == 301 or r.status_code == 302 or r.status_code == 307 or r.status_code == 308:
        # Perform URL redirect
        print('EDP-GW authentication HTTP code:', r.status_code, r.reason)
        new_host = r.headers['Location']
        if new_host is not None:
            print('Perform URL redirect to ', new_host)
            return get_sts_token(current_refresh_token, new_host)
        return None, None, None
    elif r.status_code == 400 or r.status_code == 401:
        # Retry with username and password
        print('EDP-GW authentication HTTP code:', r.status_code, r.reason)
        if current_refresh_token:
            # Refresh token may have expired. Try using our password.
            print('Retry with username and password')
            return get_sts_token(None)
        return None, None, None
    elif r.status_code == 403 or r.status_code == 451:
        # Stop retrying with the request
        print('EDP-GW authentication HTTP code:', r.status_code, r.reason)
        print('Stop retrying with the request')
        return None, None, None
    else:
        # Retry the request to the API gateway
        print('EDP-GW authentication HTTP code:', r.status_code, r.reason)
        print('Retry the request to the API gateway')
        return get_sts_token(current_refresh_token)

if __name__ == "__main__":
    # Get command line parameters
    try:
        opts, args = getopt.getopt(sys.argv[1:], "", ["help", "hostname=", "port=", "app_id=", "user=", "clientid=", "password=",
                                                      "position=", "auth_url=", "scope=", "ric=", "service=", "view=", "no_streaming"])
    except getopt.GetoptError:
        print('Usage: market_price_edpgw_authentication.py [--hostname hostname] [--port port] [--app_id app_id] '
              '[--user user] [--clientid clientid] [--password password] [--position position] [--auth_url auth_url] '
              '[--scope scope] [--ric ric] [--service service] [--view view] [--no_streaming] [--help]')
        sys.exit(2)
    for opt, arg in opts:
        if opt in "--help":
            print('Usage: market_price_edpgw_authentication.py [--hostname hostname] [--port port] [--app_id app_id] '
                  '[--user user] [--clientid clientid] [--password password] [--position position] [--auth_url auth_url] '
                  '[--scope scope] [--ric ric] [--service service] [--view view] [--no_streaming] [--help]')
            sys.exit(0)
        elif opt in "--hostname":
            hostname = arg
        elif opt in "--port":
            port = arg
        elif opt in "--app_id":
            app_id = arg
        elif opt in "--user":
            user = arg
        elif opt in "--clientid":
            clientid = arg
        elif opt in "--password":
            password = arg
        elif opt in "--position":
            position = arg
        elif opt in "--auth_url":
            auth_url = arg
        elif opt in "--scope":
            scope = arg
        elif opt in "--ric":
            ric = arg
        elif opt in "--service":
            service = arg
        elif opt in "--view":
            view = arg.split(',')
        elif opt in "--no_streaming":
            no_streaming = True

    if user == '' or password == '' or  hostname == '' or clientid == '':
        print("user, clientid, password, and hostname are required options")
        sys.exit(2)

    if position == '':
        # Populate position if possible
        try:
            position_host = socket.gethostname()
            position = socket.gethostbyname(position_host) + "/" + position_host
        except socket.gaierror:
            position = "127.0.0.1/net"

    sts_token, refresh_token, expire_time = get_sts_token(None)
    if not sts_token:
        sys.exit(1)

    # Start websocket handshake
    ws_address = "wss://{}:{}/WebSocket".format(hostname, port)
    print("Connecting to WebSocket " + ws_address + " ...")
    web_socket_app = websocket.WebSocketApp(ws_address, on_message=on_message,
                                            on_error=on_error,
                                            on_close=on_close,
                                            subprotocols=['tr_json2'])
    web_socket_app.on_open = on_open

    # Event loop
    wst = threading.Thread(target=web_socket_app.run_forever, kwargs={'sslopt': {'check_hostname': False}})
    wst.start()

    try:
        while True:
            # Give 30 seconds to obtain the new security token and send reissue
            if int(expire_time) > 30:
                time.sleep(int(expire_time) - 30)
            else:
                # Fail the refresh since value too small
                sys.exit(1)
            sts_token, refresh_token, expire_time = get_sts_token(refresh_token)
            if not sts_token:
                sys.exit(1)

            # Update token.
            if logged_in:
                send_login_request(sts_token, True)
    except KeyboardInterrupt:
        web_socket_app.close()

File: Applications/Examples/EDP/java/MarketPriceEdpGwAuthentication.java
//|-----------------------------------------------------------------------------
//|            This source code is provided under the Apache 2.0 license      --
//|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
//|                See the project's LICENSE.md for details.                  --
//|            Copyright (C) 2019-2020 Refinitiv. All rights reserved.        --
//|-----------------------------------------------------------------------------


import com.neovisionaries.ws.client.WebSocket;
import com.neovisionaries.ws.client.WebSocketAdapter;
import com.neovisionaries.ws.client.WebSocketException;
import com.neovisionaries.ws.client.WebSocketExtension;
import com.neovisionaries.ws.client.WebSocketFactory;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.net.Inet4Address;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Properties;

import org.json.*;
import org.apache.commons.cli.*;
import org.apache.http.*;
import org.apache.http.client.HttpClient;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.params.ClientPNames;
import org.apache.http.params.HttpParams;
import org.apache.http.params.BasicHttpParams;
import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.ssl.SSLContextBuilder;
import org.apache.http.util.EntityUtils;

/**
 * This example demonstrates authenticating via the Elektron Real-Time Service and Elektron Data Platform Gateway, and logging in with the retrieved token to retrieve market content.
 * It does so by:
 * - Authenticating via HTTP Post request to the Gateway
 * - Opening a WebSocket to the Elektron Real-Time Service
 * - Logging into the Real-Time Service using the token retrieved from the Gateway
 * - Requesting market content
 * - Perodically re-authenticating to the Gateway, and providing the updated token to the Real-Time Service.
 */
public class MarketPriceEdpGwAuthentication {

    public static String server;
    public static String hostname = "127.0.0.1";
    public static String port = "443";
    public static String user = "root";
    public static String clientid = "";
    public static String position = "";
    public static String appId = "256";
    public static WebSocket ws = null;
    public static String authToken = "";
    public static String password = "";
    public static String newPassword = "";
    public static String authUrl = "https://api.refinitiv.com:443/auth/oauth2/v1/token";
    public static String ric = "/TRI.N";
    public static String service = "ELEKTRON_DD";
    public static String scope = "trapi";
    public static JSONObject authJson = null;

    final private static int passwordLengthMask               = 0x1;
    final private static int passwordUppercaseLetterMask      = 0x2;
    final private static int passwordLowercaseLetterMask      = 0x4;
    final private static int passwordDigitMask                = 0x8;
    final private static int passwordSpecialCharacterMask     = 0x10;
    final private static int passwordInvalidCharacterMask     = 0x20;


    // Default password policy
    final private static int passwordLengthMin                = 30;
    final private static int passwordUppercaseLetterMin       = 1;
    final private static int passwordLowercaseLetterMin       = 1;
    final private static int passwordDigitMin                 = 1;
    final private static int passwordSpecialCharacterMin      = 1;
    final private static String passwordSpecialCharacterSet   = "~!@#$%^&*()-_=+[]{}|;:,.<>/?";
    final private static int passwordMinNumberOfCategories    = 3;
    
    public static void main(String[] args) {

    Options options = new Options();

        options.addOption(Option.builder().longOpt("hostname").required().hasArg().desc("hostname").build());
        options.addOption(Option.builder().longOpt("port").hasArg().desc("port").build());
        options.addOption(Option.builder().longOpt("app_id").hasArg().desc("app_id").build());
        options.addOption(Option.builder().longOpt("user").required().hasArg().desc("user").build());
        options.addOption(Option.builder().longOpt("clientid").required().hasArg().desc("clientid").build());
        options.addOption(Option.builder().longOpt("position").hasArg().desc("position").build());
        options.addOption(Option.builder().longOpt("password").required().hasArg().desc("password").build());
        options.addOption(Option.builder().longOpt("newPassword").hasArg().desc("newPassword").build());
        options.addOption(Option.builder().longOpt("auth_url").hasArg().desc("auth_url").build());
        options.addOption(Option.builder().longOpt("ric").hasArg().desc("ric").build());
        options.addOption(Option.builder().longOpt("service").hasArg().desc("service").build());
        options.addOption(Option.builder().longOpt("scope").hasArg().desc("scope").build());
        options.addOption(Option.builder().longOpt("help").desc("help").build());


        CommandLineParser parser = new DefaultParser();
        HelpFormatter formatter = new HelpFormatter();
        CommandLine cmd;

        try {
            cmd = parser.parse(options, args);
        } catch (org.apache.commons.cli.ParseException e) {
            System.out.println(e.getMessage());
            formatter.printHelp("MarketPriceEdpGwAuthentication", options);
            System.exit(1);
            return;
        }

        if(cmd.hasOption("help"))
        {
             formatter.printHelp("MarketPriceEdpGwAuthentication", options);
             System.exit(0);
        }
        if(cmd.hasOption("hostname"))
            hostname = cmd.getOptionValue("hostname");
        if(cmd.hasOption("port"))
            port = cmd.getOptionValue("port");
        if(cmd.hasOption("app_id"))
            appId = cmd.getOptionValue("app_id");
        if(cmd.hasOption("user"))
            user = cmd.getOptionValue("user");
        if(cmd.hasOption("clientid"))
            clientid = cmd.getOptionValue("clientid");
        if(cmd.hasOption("password"))
            password = cmd.getOptionValue("password");
        if(cmd.hasOption("auth_url"))
            authUrl = cmd.getOptionValue("auth_url");
        if(cmd.hasOption("position"))
        {
            position = cmd.getOptionValue("position");
        }
        else
        {
            try {
                position = Inet4Address.getLocalHost().getHostAddress();
            } catch (UnknownHostException e) {
                // If localhost isn't defined, use 127.0.0.1.
                position = "127.0.0.1/net";
            }
        }
        if(cmd.hasOption("ric"))
            ric = cmd.getOptionValue("ric");
        if(cmd.hasOption("service"))
            service = cmd.getOptionValue("service");
        if(cmd.hasOption("scope"))
            scope = cmd.getOptionValue("scope");
        if(cmd.hasOption("newPassword")) {
        	newPassword = cmd.getOptionValue("newPassword");
        	if ((newPassword == null) || (newPassword.length() == 0)) {
        		System.out.println("Value of the option newPassword cannot be empty");
        		System.exit(1);
        	}
        	
        	int result = checkPassword(newPassword);
        	if ((result & passwordInvalidCharacterMask) != 0) {
        		System.out.println("New password contains invalid symbol(s)");
        		System.out.println("Valid symbols are [A-Z][a-z][0-9]" + passwordSpecialCharacterSet);
        		System.exit(0);
        	}
        	
        	if ((result & passwordLengthMask) != 0) {
        		System.out.println("New password length should be at least " 
        	            + passwordLengthMin
        	            + " characters");
        		System.exit(0);
        	}
        	int countCategories = 0;
        	for (int mask = passwordUppercaseLetterMask; mask <= passwordSpecialCharacterMask; mask <<= 1) {
        		if ((result & mask) == 0) {
        			countCategories++;
        		}
        	}
        	if (countCategories < passwordMinNumberOfCategories) {
        		System.out.println("Password must contain characters belonging to at least three of the following four categories:\n"
	    				+ "uppercase letters, lovercase letters, digits, and special characters.\n");
        		System.exit(0);
        	}
         	if (!changePassword(authUrl)) {
         		System.exit(0);
         	}
         	password = newPassword;
         	newPassword = "";
        }

        try {

            // Connect to the gateway and authenticate (using our username and password)
            authJson = getAuthenticationInfo(null);
            if (authJson == null)
                System.exit(1);

            // Determine when the access token expires. We will re-authenticate before then.
            int expireTime = Integer.parseInt(authJson.getString("expires_in"));

            server = String.format("wss://%s:%s/WebSocket", hostname, port);
            System.out.println("Connecting to WebSocket " + server + " ...");
            ws = connect();

            while(true) {
                // Continue using current token until 90% of initial time before it expires.
                Thread.sleep(expireTime * 900);  // The value 900 means 90% of expireTime in milliseconds

                // Connect to the gateway and re-authenticate, using the refresh token provided in the previous response
                authJson = getAuthenticationInfo(authJson);
                if (authJson == null)
                    System.exit(1);

                // If expiration time returned by refresh request is less then initial expiration time,
                // re-authenticate using password
                int refreshingExpireTime = Integer.parseInt(authJson.getString("expires_in"));
                if (refreshingExpireTime != expireTime) {
                	System.out.println("expire time changed from " + expireTime + " sec to " + refreshingExpireTime + 
                    		" sec; retry with password");
                    authJson = getAuthenticationInfo(null);
                    if (authJson == null)
                        System.exit(1);
                    expireTime = Integer.parseInt(authJson.getString("expires_in"));
                }

                // Send the updated access token over the WebSocket.
                sendLoginRequest(ws, authJson.getString("access_token"), false);

            }
        } catch (IOException e) {
            e.printStackTrace();
        } catch (WebSocketException e) {
            e.printStackTrace();
        } catch (JSONException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
    
    public static int checkPassword(String pwd) {
    	int result = 0;
    	
    	if (pwd.length() < passwordLengthMin) {
    		result |= passwordLengthMask;
    	} 
    	
    	int countUpper = 0;
    	int countLower = 0;
        int countDigit = 0;
        int countSpecial = 0;
        
        for (int i = 0; i < pwd.length(); i++) {
        	char c = pwd.charAt(i);
        	StringBuffer currentSymbol = new StringBuffer(1);
        	currentSymbol.append(c);

        	String charAsString = new String(currentSymbol);
        	if ((!charAsString.matches("[A-Za-z0-9]")) && (!passwordSpecialCharacterSet.contains(currentSymbol))) {
        		result |= passwordInvalidCharacterMask;
        	}
        	
        	if (Character.isUpperCase(c)) {
        		countUpper++;
        	}
        	if (Character.isLowerCase(c)) {
        		countLower++;
        	}
        	if (Character.isDigit(c)) {
        		countDigit++;
        	}        	
        	if (passwordSpecialCharacterSet.contains(currentSymbol))  {
        		countSpecial++;
        	}
        }
        
        if (countUpper < passwordUppercaseLetterMin) {
        	result |= passwordUppercaseLetterMask;
        }
        if (countLower < passwordLowercaseLetterMin) {
        	result |= passwordLowercaseLetterMask;
        }
        if (countDigit < passwordDigitMin) {
        	result |= passwordDigitMask;
        }
        if (countSpecial < passwordSpecialCharacterMin) {
        	result |= passwordSpecialCharacterMask;
        }
        
    	return result;
    }
    
    /**
     * Send a request to change password and receive an answer.
     */   
    public static boolean changePassword(String authServer) {
    	boolean result = false;
        try
        {
            SSLConnectionSocketFactory sslsf = new SSLConnectionSocketFactory(new SSLContextBuilder().build());

            HttpClient httpclient = HttpClients.custom().setSSLSocketFactory(sslsf).build();
            HttpPost httppost = new HttpPost(authServer);
            HttpParams httpParams = new BasicHttpParams();

            // Disable redirect
            httpParams.setParameter(ClientPNames.HANDLE_REDIRECTS, false);

            // Set request parameters.
            List<NameValuePair> params = new ArrayList<NameValuePair>(2);
            params.add(new BasicNameValuePair("client_id", clientid));
            params.add(new BasicNameValuePair("username", user));
            params.add(new BasicNameValuePair("grant_type", "password"));
            params.add(new BasicNameValuePair("password", password));
            params.add(new BasicNameValuePair("newPassword", newPassword));
            params.add(new BasicNameValuePair("scope", scope));
            params.add(new BasicNameValuePair("takeExclusiveSignOnControl", "true"));
            System.out.println("Sending password change request to " + authUrl);

            httppost.setParams(httpParams);
            httppost.setEntity(new UrlEncodedFormEntity(params, "UTF-8"));

            //Execute and get the response.
            HttpResponse response = httpclient.execute(httppost);

            int statusCode = response.getStatusLine().getStatusCode();

            JSONObject responseJson = new JSONObject(EntityUtils.toString(response.getEntity()));
            switch ( statusCode ) {
            case HttpStatus.SC_OK:                  // 200
                // Password change was successful.
                System.out.println("Password was successfully changed:");
                result = true;
                break;

            case HttpStatus.SC_MOVED_PERMANENTLY:              // 301
            case HttpStatus.SC_MOVED_TEMPORARILY:              // 302
            case HttpStatus.SC_TEMPORARY_REDIRECT:             // 307
            case 308:                                          // 308 HttpStatus.SC_PERMANENT_REDIRECT
                // Perform URL redirect
                System.out.println("Password change HTTP code: " + response.getStatusLine().getStatusCode() + " " + response.getStatusLine().getReasonPhrase());
                Header header = response.getFirstHeader("Location");
                if( header != null )
                {
                    String newHost = header.getValue();
                    if ( newHost != null )
                    {
                        System.out.println("Perform URL redirect to " + newHost);
                        result = changePassword(newHost);
                    } else {
                    	result = false;
                    }
                }
                break;
            default:
                // Error 4XX or 5XX
                System.out.println("Password change failure\n" 
                		+ response.getStatusLine().getStatusCode() + " " 
                		+ response.getStatusLine().getReasonPhrase());
                System.out.println(responseJson.toString(2));
                result = false;
            }
        } catch (Exception e) {
            System.out.println("Password change failure:");
            e.printStackTrace();
            result = false;
        }
        return result;
    }
    
    
    /**
     * Connect to the Realtime Service over a WebSocket.
     */
    public static WebSocket connect() throws IOException, WebSocketException
    {
        return new WebSocketFactory()
                .createSocket(server)
                .addProtocol("tr_json2")
                .addListener(new WebSocketAdapter() {

                    /**
                     * Called when message received, parse message into JSON for processing
                     */
                    public void onTextMessage(WebSocket websocket, String message) throws JSONException {
                        if(!message.isEmpty()) {
                            System.out.println("RECEIVED:");

                            JSONArray jsonArray = new JSONArray(message);

                            System.out.println(jsonArray.toString(2));

                            for (int i = 0; i < jsonArray.length(); ++i)
                                processMessage(websocket, jsonArray.getJSONObject(i));
                        }
                    }

                    /**
                     * Called when handshake is complete and websocket is open, send login
                     */
                    public void onConnected(WebSocket websocket, Map<String, List<String>> headers) throws JSONException {
                        System.out.println("WebSocket successfully connected!");
                        sendLoginRequest(websocket, authJson.getString("access_token"), true);
                    }
                })
                .addExtension(WebSocketExtension.PERMESSAGE_DEFLATE)
                .connect();
    }

    /**
     * Authenticate to the gateway via an HTTP post request.
     * Initially authenticates using the specified password. If information from a previous authentication response is provided, it instead authenticates using
     * the refresh token from that response. Uses authUrl as url.
     * @param previousAuthResponseJson Information from a previous authentication, if available
     * @return A JSONObject containing the authentication information from the response.
     */
    public static JSONObject getAuthenticationInfo(JSONObject previousAuthResponseJson) {
        String url = authUrl;
        return getAuthenticationInfo(previousAuthResponseJson, url);
    }

    /**
     * Authenticate to the gateway via an HTTP post request.
     * Initially authenticates using the specified password. If information from a previous authentication response is provided, it instead authenticates using
     * the refresh token from that response.
     * @param previousAuthResponseJson Information from a previous authentication, if available
     * @param url The HTTP post url
     * @return A JSONObject containing the authentication information from the response.
     */
    public static JSONObject getAuthenticationInfo(JSONObject previousAuthResponseJson, String url) {
        try
        {
            SSLConnectionSocketFactory sslsf = new SSLConnectionSocketFactory(new SSLContextBuilder().build());

            HttpClient httpclient = HttpClients.custom().setSSLSocketFactory(sslsf).build();
            HttpPost httppost = new HttpPost(url);
            HttpParams httpParams = new BasicHttpParams();

            // Disable redirect
            httpParams.setParameter(ClientPNames.HANDLE_REDIRECTS, false);

            // Set request parameters.
            List<NameValuePair> params = new ArrayList<NameValuePair>(2);
            params.add(new BasicNameValuePair("client_id", clientid));
            params.add(new BasicNameValuePair("username", user));

            if (previousAuthResponseJson == null)
            {
                // First time through, send password.
                params.add(new BasicNameValuePair("grant_type", "password"));
                params.add(new BasicNameValuePair("password", password));
                params.add(new BasicNameValuePair("scope", scope));
                params.add(new BasicNameValuePair("takeExclusiveSignOnControl", "true"));
                System.out.println("Sending authentication request with password to " + url + "...");

            }
            else
            {
                // Use the refresh token we got from the last authentication response.
                params.add(new BasicNameValuePair("grant_type", "refresh_token"));
                params.add(new BasicNameValuePair("refresh_token", previousAuthResponseJson.getString("refresh_token")));
                System.out.println("Sending authentication request with refresh token to " + url + "...");
            }

            httppost.setParams(httpParams);
            httppost.setEntity(new UrlEncodedFormEntity(params, "UTF-8"));

            //Execute and get the response.
            HttpResponse response = httpclient.execute(httppost);

            int statusCode = response.getStatusLine().getStatusCode();

            switch ( statusCode ) {
            case HttpStatus.SC_OK:                  // 200
                // Authentication was successful. Deserialize the response and return it.
                JSONObject responseJson = new JSONObject(EntityUtils.toString(response.getEntity()));
                System.out.println("EDP-GW Authentication succeeded. RECEIVED:");
                System.out.println(responseJson.toString(2));
                return responseJson;
            case HttpStatus.SC_MOVED_PERMANENTLY:              // 301
            case HttpStatus.SC_MOVED_TEMPORARILY:              // 302
            case HttpStatus.SC_TEMPORARY_REDIRECT:             // 307
            case 308:                                          // 308 HttpStatus.SC_PERMANENT_REDIRECT
                // Perform URL redirect
                System.out.println("EDP-GW authentication HTTP code: " + response.getStatusLine().getStatusCode() + " " + response.getStatusLine().getReasonPhrase());
                Header header = response.getFirstHeader("Location");
                if( header != null )
                {
                    String newHost = header.getValue();
                    if ( newHost != null )
                    {
                        System.out.println("Perform URL redirect to " + newHost);
                        return getAuthenticationInfo(previousAuthResponseJson, newHost);
                    }
                }
                return null;
            case HttpStatus.SC_BAD_REQUEST:                    // 400
            case HttpStatus.SC_UNAUTHORIZED:                   // 401
                // Retry with username and password
                System.out.println("EDP-GW authentication HTTP code: " + response.getStatusLine().getStatusCode() + " " + response.getStatusLine().getReasonPhrase());
                if (previousAuthResponseJson != null)
                {
                    System.out.println("Retry with username and password");
                    return getAuthenticationInfo(null);
                }
                return null;
            case HttpStatus.SC_FORBIDDEN:                      // 403
            case 451:                                          // 451 Unavailable For Legal Reasons
                // Stop retrying with the request
                System.out.println("EDP-GW authentication HTTP code: " + response.getStatusLine().getStatusCode() + " " + response.getStatusLine().getReasonPhrase());
                System.out.println("Stop retrying with the request");
                return null;
            default:
                // Retry the request to the API gateway
                System.out.println("EDP-GW authentication HTTP code: " + response.getStatusLine().getStatusCode() + " " + response.getStatusLine().getReasonPhrase());
                System.out.println("Retry the request to the API gateway");
                return getAuthenticationInfo(previousAuthResponseJson);
            }
        } catch (Exception e) {
            System.out.println("EDP-GW authentication failure:");
            e.printStackTrace();
            return null;
        }
    }

    /**
     * Generate a login request from command line data (or defaults) and send
     * Used for both the initial login and subsequent logins that send updated access tokens.
     * @param websocket Websocket to send the request on
     * @param authToken Token to use
     * @param isFirstLogin Whether this is our first login
     * @throws JSONException
     */
    public static void sendLoginRequest(WebSocket websocket, String authToken, boolean isFirstLogin) throws JSONException {
        String loginJsonString = "{\"ID\":1,\"Domain\":\"Login\",\"Key\":{\"Elements\":{\"ApplicationId\":\"\",\"Position\":\"\",\"AuthenticationToken\":\"\"},\"NameType\":\"AuthnToken\"}}";
        JSONObject loginJson = new JSONObject(loginJsonString);
        loginJson.getJSONObject("Key").getJSONObject("Elements").put("AuthenticationToken", authToken);
        loginJson.getJSONObject("Key").getJSONObject("Elements").put("ApplicationId", appId);
        loginJson.getJSONObject("Key").getJSONObject("Elements").put("Position", position);

        if (!isFirstLogin) // If this isn't our first login, we don't need another refresh for it.
            loginJson.put("Refresh", false);

        websocket.sendText(loginJson.toString());
        System.out.println("SENT:\n" + loginJson.toString(2));
    }

    /**
     * Process a message received over the WebSocket
     * @param websocket Websocket the message was received on
     * @param messageJson Deserialized JSON message
     * @throws JSONException
     */
    public static void processMessage(WebSocket websocket, JSONObject messageJson) throws JSONException {
        String messageType = messageJson.getString("Type");

        switch(messageType)
        {
            case "Refresh":
            case "Status":
                if(messageJson.has("Domain")) {
                    String messageDomain = messageJson.getString("Domain");
                    if(messageDomain.equals("Login")) {
                        // Check message state to see if login succeeded. If so, send item request. Otherwise stop.
                        JSONObject messageState = messageJson.optJSONObject("State");
                        if (messageState != null)
                        {
                            if (!messageState.getString("Stream").equals("Open") || !messageState.getString("Data").equals("Ok"))
                            {
                                System.out.println("Login failed.");
                                System.exit(1);
                            }

                            // Login succeeded, send item request.
                            sendRequest(websocket);
                        }

                    }
                }
                break;

            case "Ping":
                String pongJsonString = "{\"Type\":\"Pong\"}";
                JSONObject pongJson = new JSONObject(pongJsonString);
                websocket.sendText(pongJsonString);
                System.out.println("SENT:\n" + pongJson.toString(2));
                break;
            default:
                break;
        }
    }

    /**
     * Create and send simple Market Price request
     * @param websocket Websocket to send the message on
     * @throws JSONException
     */
    public static void sendRequest(WebSocket websocket) throws JSONException {
        String requestJsonString;
        requestJsonString = "{\"ID\":2,\"Key\":{\"Name\":\"" + ric + "\",\"Service\":\"" + service + "\"}}";
        JSONObject mpRequestJson = new JSONObject(requestJsonString);
        websocket.sendText(requestJsonString);
        System.out.println("SENT:\n" + mpRequestJson.toString(2));
    }
}
